src/core/qgsvectorlayer.cpp

```diff
@@ -944,6 +944,20 @@ bool QgsVectorLayer::addFeature( QgsFeature &feature, Flags )
   if ( !mValid || !mEditBuffer || !mDataProvider )
     return false;
 
+
+  if ( mGeometryOptions.geometryPrecision != 0.0 || mGeometryOptions.removeDuplicateNodes )
+  {
+    QgsGeometry geom = feature.geometry();
+
+    if ( mGeometryOptions.geometryPrecision != 0.0 )
```

Comment:
Can you move these operations to a common private function to avoid the code duplication?

Reply:
Done.